dotnet/src/Connectors/Connectors.Memory.SqlServer/SqlServerCommandBuilder.cs

```diff
@@ -0,0 +1,547 @@
+        }
+
+        if (any)
+        {
+            --sb.Length; // remove the last comma
+        }
+
+        return sb;
+    }
+
+    private static StringBuilder AppendKeyParameterList<TKey>(this StringBuilder sb,
+        IEnumerable<TKey> keys, SqlCommand command, VectorStoreRecordKeyProperty keyProperty, out bool emptyKeys)
+    {
+        int keyIndex = 0;
+        foreach (TKey key in keys)
+        {
+            // The caller ensures that keys collection is not null.
+            // We need to ensure that none of the keys is null.
+            Verify.NotNull(key);
+
+            sb.AppendParameterName(keyProperty, ref keyIndex, out string keyParamName);
+            sb.Append(',');
+            command.AddParameter(keyProperty, keyParamName, key);
+        }
+
+        emptyKeys = keyIndex == 0;
+        sb.Length--; // remove the last comma
+        return sb;
+    }
+
+    private static SqlCommand CreateCommand(this SqlConnection connection, StringBuilder sb)
+    {
+        SqlCommand command = connection.CreateCommand();
+        command.CommandText = sb.ToString();
+        return command;
+    }
+
+    private static void AddParameter(this SqlCommand command, VectorStoreRecordProperty property, string name, object? value)
+    {
+        switch (value)
+        {
+            case null when property.PropertyType == typeof(byte[]):
+                command.Parameters.Add(name, System.Data.SqlDbType.VarBinary).Value = DBNull.Value;
+                break;
+            case null:
+            case ReadOnlyMemory<float> vector when vector.Length == 0:
+                command.Parameters.AddWithValue(name, DBNull.Value);
+                break;
+            case byte[] buffer:
+                command.Parameters.Add(name, System.Data.SqlDbType.VarBinary).Value = buffer;
+                break;
+            case ReadOnlyMemory<float> vector:
+                command.Parameters.AddWithValue(name, JsonSerializer.Serialize(vector));
+                break;
+            default:
+                command.Parameters.AddWithValue(name, value);
+                break;
+        }
+    }
+
+    private static (string sqlName, string? autoGenerate) Map(Type type)
+    {
+        return type switch
+        {
+            Type t when t == typeof(byte) => ("TINYINT", null),
+            Type t when t == typeof(short) => ("SMALLINT", null),
+            Type t when t == typeof(int) => ("INT", "IDENTITY(1,1)"),
+            Type t when t == typeof(long) => ("BIGINT", "IDENTITY(1,1)"),
+            Type t when t == typeof(Guid) => ("UNIQUEIDENTIFIER", "DEFAULT NEWSEQUENTIALID()"),
+            Type t when t == typeof(string) => ("NVARCHAR(255)", null),
+            Type t when t == typeof(byte[]) => ("VARBINARY(MAX)", null),
+            Type t when t == typeof(bool) => ("BIT", null),
+            Type t when t == typeof(DateTime) => ("DATETIME2", null),
+#if NET
+            Type t when t == typeof(TimeOnly) => ("TIME", null),
```

Comment:
We also need DateOnly here, mapped to DATE.

Reply:
I took a look at implementing `DateOnly` and I don't see a way to easily do that as of now. The reason is that the `SqlDataReader` maps such an output as `DateTime` and `reader.GetProviderSpecificFieldType(this._sqlDataReader.GetOrdinal(storageName))` also returns `DateTime` and we also support `DATETIME2` which does the same.

For `TIME` it was similar, as it also returns `TimeSpan` rather than `TimeOnly`. But we don't support `TimeSpan`, so we can just map all `TimeSpan` instances into `TimeOnly`.

I know I could give the reader the property infos or the definitions to deal with that, I just don't believe it's worth it as of now, especially if we want to get rid of this layer (the need to represent `SqlDataReader` as `IDictionary<string, object?>`).